lib/sudo-common/src/resolve.rs

```diff
@@ -0,0 +1,209 @@
+    Id(T),
+}
+
+impl<'a, T: FromStr> NameOrId<'a, T> {
+    pub fn parse(input: &'a str) -> Option<Self> {
+        if input.is_empty() {
+            None
+        } else if let Some(stripped) = input.strip_prefix('#') {
+            stripped.parse::<T>().ok().map(|id| Self::Id(id))
+        } else {
+            Some(Self::Name(input))
+        }
+    }
+}
+
+pub(crate) fn resolve_current_user() -> Result<User, Error> {
+    User::real()?.ok_or(Error::UserNotFound("current user".to_string()))
+}
+
+pub(crate) fn resolve_target_user(target_name_or_id: &Option<String>) -> Result<User, Error> {
+    let is_default = target_name_or_id.is_none();
+    let target_name_or_id = target_name_or_id.as_deref().unwrap_or("root");
+
+    let mut user = match NameOrId::parse(target_name_or_id) {
+        Some(NameOrId::Name(name)) => User::from_name(name)?,
+        Some(NameOrId::Id(uid)) => User::from_uid(uid)?,
+        _ => None,
+    }
+    .ok_or_else(|| Error::UserNotFound(target_name_or_id.to_string()))?;
+
+    user.is_default = is_default;
+
+    Ok(user)
+}
+
+pub(crate) fn resolve_target_group(
+    target_name_or_id: &Option<String>,
+    target_user: &User,
+) -> Result<Group, Error> {
+    match target_name_or_id.as_deref() {
+        Some(name_or_id) => match NameOrId::parse(name_or_id) {
+            Some(NameOrId::Name(name)) => Group::from_name(name)?,
+            Some(NameOrId::Id(gid)) => Group::from_gid(gid)?,
+            _ => None,
+        },
+        None => Group::from_gid(target_user.gid)?,
+    }
+    .ok_or(Error::GroupNotFound(
+        target_name_or_id
+            .clone()
+            .unwrap_or_else(|| target_user.gid.to_string()),
+    ))
+}
+
+/// Check whether a path points to a regular file and any executable flag is set
+fn is_valid_executable(path: &PathBuf) -> bool {
+    if path.is_file() {
+        match fs::metadata(path) {
+            Ok(meta) => meta.mode() & 0o111 != 0,
+            _ => false,
+        }
+    } else {
+        false
+    }
+}
+
+/// Resolve a executable name based in the PATH environment variable
+/// When resolving a path, this code checks whether the target file is
+/// a regular file and has any executable bits set. It does not specifically
+/// check for user, group, or others' executable bit.
+pub(crate) fn resolve_path(command: &Path, path: &str) -> Option<PathBuf> {
+    // To prevent command spoofing, sudo checks "." and "" (both denoting current directory)
+    // last when searching for a command in the user's PATH (if one or both are in the PATH).
+    // Depending on the security policy, the user's PATH environment variable may be modified,
+    // replaced, or passed unchanged to the program that sudo executes.
+    let mut resolve_current_path = false;
+
+    path.split(':')
+        .into_iter()
+        // register whether to look in the current directory, but first check the other PATH segments
+        .filter(|&path| {
+            if path.is_empty() || path == "." {
```

Comment:
Note: if we are going to check for ".", we should also check "..", otherwise we can still spoof commands. 

Original sudo also doesn't check for ".." by the way.